streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java

```diff
@@ -0,0 +1,334 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.internals.RecordContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * An in-memory LRU cache store similar to {@link MemoryLRUCache} but byte-based, not
+ * record based
+ *
+ * @see org.apache.kafka.streams.state.Stores#create(String)
+ */
+public class ThreadCache {
+    private static final Logger log = LoggerFactory.getLogger(ThreadCache.class);
+    private final String name;
+    private final long maxCacheSizeBytes;
+    private final Map<String, NamedCache> caches = new HashMap<>();
+    private final ThreadCacheMetrics metrics;
+
+    // JMX stats
+    private Sensor puts = null;
```

Comment:
Do we need metrics for the global thread cache here? I felt the NamedCache's metrics are sufficient. Main rationale is to `rather regret and add later, than regret and remove later`.